File: languages/c/k-means-par.c
// k-means-par: parallel (wall-clock scaling track) Lloyd's clustering. Same result
// as serial k-means.c for ANY core count. Invocation: k-means-par <cores> <n>.
// The ASSIGNMENT step (the O(N*K*D) hot loop) is parallelised over point-bands:
// worker w computes assign[i] for points [w*N/cores, (w+1)*N/cores), reading the
// read-only centroids. The centroid UPDATE (ssum/cnt accumulation + floor-mean +
// empty-cluster-unchanged) runs SERIALLY after the join, exactly as serial. The
// strict-< lowest-index tie-break is preserved (each worker scans k=0..K-1 in order).
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include <time.h>

static long long now_ns(void) { struct timespec t; clock_gettime(CLOCK_MONOTONIC, &t); return (long long)t.tv_sec * 1000000000LL + t.tv_nsec; }

#define P 1000000007L
#define K 16
#define D 4
#define ITERS 10
#define RANGE 256

static long lcg(long s) { return (s * 1103515245L + 12345L) & 0x7fffffffL; }

static int N;
static long *pt;
static long cen[K * D];
static int *assign;

typedef struct { int lo, hi; } Band;

static void *assign_worker(void *arg) {
    Band *b = (Band *)arg;
    for (int i = b->lo; i < b->hi; i++) {
        int best = 0; long bd = -1;
        for (int k = 0; k < K; k++) {
            long dist = 0;
            for (int d = 0; d < D; d++) { long df = pt[(long)i * D + d] - cen[k * D + d]; dist += df * df; }
            if (bd < 0 || dist < bd) { bd = dist; best = k; }
        }
        assign[i] = best;
    }
    return NULL;
}

static void parallel_assign(int cores) {
    pthread_t th[256];
    Band bands[256];
    for (int w = 0; w < cores; w++) {
        bands[w].lo = (int)((long)w * N / cores);
        bands[w].hi = (int)((long)(w + 1) * N / cores);
        pthread_create(&th[w], NULL, assign_worker, &bands[w]);
    }
    for (int w = 0; w < cores; w++) pthread_join(th[w], NULL);
}

int main(int argc, char **argv) {
    int cores = argc > 1 ? atoi(argv[1]) : 1;
    N = argc > 2 ? atoi(argv[2]) : 8000;
    if (cores < 1) cores = 1;
    if (cores > N) cores = N;
    if (cores > 256) cores = 256;

    pt = malloc((size_t)N * D * sizeof(long));
    long s = 42;
    for (long i = 0; i < (long)N * D; i++) { s = lcg(s); pt[i] = s % RANGE; }
    for (int i = 0; i < K * D; i++) cen[i] = pt[i];   // initial centroids = first K points
    assign = malloc((size_t)N * sizeof(int));

    long long _t0 = now_ns();
    for (int iter = 0; iter < ITERS; iter++) {
        parallel_assign(cores);                       // parallel assignment
        long ssum[K * D] = {0}; long cnt[K] = {0};     // serial update (identical to serial)
        for (int i = 0; i < N; i++) {
            int k = assign[i]; cnt[k]++;
            for (int d = 0; d < D; d++) ssum[k * D + d] += pt[(long)i * D + d];
        }
        for (int k = 0; k < K; k++)
            if (cnt[k] > 0)
                for (int d = 0; d < D; d++) cen[k * D + d] = ssum[k * D + d] / cnt[k];
    }

    parallel_assign(cores);                            // final assignment with final centroids
    fprintf(stderr, "COMPUTE_NS %lld\n", now_ns() - _t0);

    long h = 0;
    for (int i = 0; i < K * D; i++) h = (h * 31 + cen[i]) % P;
    for (int i = 0; i < N; i++) h = (h * 31 + assign[i]) % P;
    printf("%ld\n", h);
    printf("k-means(%d)\n", N);

    free(pt); free(assign);
    return 0;
}

File: languages/rust/k-means-par.rs
// k-means-par: parallel Lloyd's k-means for the wall-clock scaling track.
// Invocation: k-means-par <cores> <n>
// Output: identical to serial k-means at the same n (core-invariant).
//
// Decomposition:
//   ASSIGNMENT step: partition points into `cores` bands.  Each worker computes
//     assignments for its points (same strict-< lowest-index tie-break as serial)
//     and accumulates partial per-cluster coordinate sums and counts.
//   UPDATE step: main thread merges partial sums/counts from all workers and
//     applies floor-mean centroid update (empty-cluster unchanged), same as serial.
//
// Workers write disjoint sub-slices of `assign`.  The partial sums/counts are
// per-worker private arrays passed back via ordinary move captures + a channel-
// free collect after scope.  No atomics on the compute path.
//
// Build:
//   rustc -O languages/rust/k-means-par.rs -o /app/k-means-par
use std::env;

const P: i64 = 1000000007;
const K: usize = 16;
const D: usize = 4;
const ITERS: usize = 10;
const RANGE: i64 = 256;

fn lcg(s: i64) -> i64 {
    (s.wrapping_mul(1103515245).wrapping_add(12345)) & 0x7fffffff
}

/// Assign points [start..end) to nearest centroid (strict-< tie-break).
/// Writes into assign[start..end].
/// Returns (partial_ssum[K*D], partial_cnt[K]) for the assigned points.
fn assign_band(
    pt: &[i64],
    cen: &[i64],
    assign: &mut [i64],
    start: usize,
    end: usize,
    n: usize,
) -> (Vec<i64>, Vec<i64>) {
    let _ = n; // not needed but kept for clarity
    let mut partial_ssum = vec![0i64; K * D];
    let mut partial_cnt = vec![0i64; K];
    for i in start..end {
        let mut best: usize = 0;
        let mut bd: i64 = -1;
        for k in 0..K {
            let mut dist: i64 = 0;
            for d in 0..D {
                let df = pt[i * D + d] - cen[k * D + d];
                dist += df * df;
            }
            if bd < 0 || dist < bd {
                bd = dist;
                best = k;
            }
        }
        assign[i - start] = best as i64;
        partial_cnt[best] += 1;
        for d in 0..D {
            partial_ssum[best * D + d] += pt[i * D + d];
        }
    }
    (partial_ssum, partial_cnt)
}

fn main() {
    let mut args = env::args().skip(1);
    let cores: usize = args
        .next()
        .and_then(|s| s.parse().ok())
        .unwrap_or(1);
    let n: usize = args
        .next()
        .and_then(|s| s.parse().ok())
        .unwrap_or(8000);

    let cores = cores.max(1).min(n);

    let mut pt: Vec<i64> = vec![0; n * D];
    let mut s: i64 = 42;
    for i in 0..n * D {
        s = lcg(s);
        pt[i] = s % RANGE;
    }
    let mut cen: Vec<i64> = vec![0; K * D];
    for i in 0..K * D {
        cen[i] = pt[i];
    }
    let mut assign: Vec<i64> = vec![0; n];

    // Floor-division point bands.
    let bands: Vec<(usize, usize)> = (0..cores)
        .map(|w| (w * n / cores, (w + 1) * n / cores))
        .collect();

    let t0 = std::time::Instant::now();
    for _iter in 0..ITERS {
        // --- PARALLEL ASSIGNMENT ---
        // Each worker gets a disjoint &mut sub-slice of `assign` (via split_at_mut)
        // and RETURNS its partial (ssum, cnt) through the scoped join handle. No raw
        // pointers, no unsafe: &mut [i64] is Send and the handles preserve band order.
        let pt_ref: &[i64] = &pt;
        let cen_ref: &[i64] = &cen;
        let partials: Vec<(Vec<i64>, Vec<i64>)> = std::thread::scope(|scope| {
            let mut handles = Vec::with_capacity(cores);
            let mut remaining: &mut [i64] = &mut assign;
            let mut offset = 0usize;
            for w in 0..cores {
                let (start, end) = bands[w];
                let len = end - offset;
                let (head, tail) = remaining.split_at_mut(len);
                remaining = tail;
                offset = end;
                handles.push(scope.spawn(move || assign_band(pt_ref, cen_ref, head, start, end, n)));
            }
            handles.into_iter().map(|h| h.join().unwrap()).collect()
        });

        // --- SERIAL CENTROID UPDATE ---
        let mut ssum = vec![0i64; K * D];
        let mut cnt = vec![0i64; K];
        for (ps, pc) in &partials {
            for i in 0..K * D {
                ssum[i] += ps[i];
            }
            for k in 0..K {
                cnt[k] += pc[k];
            }
        }
        for k in 0..K {
            if cnt[k] > 0 {
                for d in 0..D {
                    cen[k * D + d] = ssum[k * D + d] / cnt[k];
                }
            }
        }
    }

    // Final assignment with final centroids (parallel, same decomposition).
    {
        let pt_ref: &[i64] = &pt;
        let cen_ref: &[i64] = &cen;
        std::thread::scope(|scope| {
            let mut remaining: &mut [i64] = &mut assign;
            let mut offset = 0usize;
            for w in 0..cores {
                let (start, end) = bands[w];
                let len = end - offset;
                let (head, tail) = remaining.split_at_mut(len);
                remaining = tail;
                offset = end;
                scope.spawn(move || {
                    for i in start..end {
                        let mut best: usize = 0;
                        let mut bd: i64 = -1;
                        for k in 0..K {
                            let mut dist: i64 = 0;
                            for d in 0..D {
                                let df = pt_ref[i * D + d] - cen_ref[k * D + d];
                                dist += df * df;
                            }
                            if bd < 0 || dist < bd {
                                bd = dist;
                                best = k;
                            }
                        }
                        head[i - start] = best as i64;
                    }
                });
            }
        });
    }
    eprintln!("COMPUTE_NS {}", t0.elapsed().as_nanos());

    // Serial checksum (identical to serial k-means).
    let mut h: i64 = 0;
    for i in 0..K * D {
        h = (h * 31 + cen[i]) % P;
    }
    for i in 0..n {
        h = (h * 31 + assign[i]) % P;
    }
    println!("{}", h);
    println!("k-means({})", n);
}
